jaxrs-api/src/main/java/jakarta/ws/rs/core/Cookie.java

```diff
@@ -216,4 +234,123 @@ public boolean equals(final Object obj) {
         }
         return true;
     }
+
+    /**
+     * JAX-RS {@link Cookie} builder class.
+     * <p>
+     * Cookie builder provides methods that let you conveniently configure and subsequently build a new
+     * {@code Cookie} instance.
+     * </p>
+     * For example:
+     *
+     * <pre>
+     * Cookie cookie = new Cookie.Builder("name")
+     *         .path("/")
+     *         .domain("domain.com")
+     *         .build();
+     * </pre>
+     * <p>
+     * It should be preferred over any {@code Cookie} telescoping constructors that have following disadvantages:
+     * <ul>
+     * <li>code not easy to read: the {@code Cookie} constructors have multiple parameters of the same type
+     * and it is difficult to determine their meaning without checking the constructor signature.</li>
+     * <li>force to pass optional parameters with {@code null} or default value even if you don't need them</li>
+     * </ul>
+     * <p>
+     *
+     * @since 3.1
+     */
+    public static class Builder extends AbstractCookieBuilder<Builder> {
+
+        public Builder(String name) {
+            super(name);
+        }
+
+        @Override
+        public Cookie build() {
+            return new Cookie(this);
+        }
+
+    }
+
+    /**
+     * JAX-RS abstract {@link Cookie} builder class.
+     *
+     * @since 3.1
+     */
+    public static abstract class AbstractCookieBuilder<T extends AbstractCookieBuilder<T>> {
```

Comment:
What benefit brings `<T>`?

Reply:
Less code duplication in implementation of `AbstractCookieBuilder`

Reply:
I don't get it. Why not simply returning `AbstractCookieBuilder` (without `<T>`) in those methods that currently return `<T>`? Maybe I am blind. Can you give an example what you _actually_ achieve here?

Reply:
This `AbstractCookieBuilder <T>` is used as a parent class of `AbstractNewCookieBuilder<T>`. 

Without the `T`, invocation of `NewCookie.Builder newCookieBuilder = new NewCookie.Builder("name").path("path")` returns an `AbstractCookieBuilder` type instead of a `NewCookie.Builder` type. So user can no longer invoke any of the specific methods (`comment`, `maxAge`...) define in `NewCookie.Builder`.

Reply:
I found this link http://www.javabyexamples.com/lets-discuss-builder-pattern about it or this one https://medium.com/@hazraarka072/fluent-builder-and-powering-it-up-with-recursive-generics-in-java-483005a85fcd.
Let me know if clearer or not.

Reply:
While this seems to be a nice programming trick to finally get in Java what you like to achieve, it is not really improve the readability or easy comprehension of the code. This originally comes from (a) `<SELF>` being much more clear than `<T>` in the conext of that blog entries, and (b) making `NewCookie` a child of `Cookie` which IMHO was no smart design right from the start of JAX-RS. Anyways, you made your case very clear, thank you, so I am OK with it once you change `<T>` to `<SELF>` to better reflect that you applied that pattern, and if you like, introducing the `private SELF self` method as in that code example.